features/251-templates/index.md

```diff
@@ -2,57 +2,92 @@
-)
+templates/phone/user_verification_message.txt    # this is an SMS message template
 ```
 
 ### Template Language Tag
 
-Each template may optionally have a language tag. In the future, we may support sending localized message. The language tag is specified in [BCP47](https://tools.ietf.org/html/bcp47).
+Each template may optionally have a language tag. The language tag is specified in [BCP47](https://tools.ietf.org/html/bcp47).
 
-### Template Key
+## Template resolution
+
+To resolve a template, the input is the template type, optionally the template key and finally the user preferred languages. The type and key is determined by the feature while the user preferred languages is provided by the user.
+
+All templates have default value so template resolution always succeed.
+
+The templates are first resolved by matching the type and the key. And then select the best language according to the user preferred languages.
+
+## Component templates
+
+Some template may depend on other templates which are included during rendering. This enables customizing a particular component of a template.
+
+For example, `auth_ui_login.html` depend on `auth_ui_header.html` and `auth_ui_footer.html` to provide the header and footer. If the developer just wants to customize the header, they do not need to provide customized templates for ALL pages. They just need to provide `auth_ui_header.html`.
+
+## Localization of the text of the template
 
-If the developer has configured `email` and `secondary_email` as login IDs. User Verification supports verifying both with different templates. Therefore, `key` is required to differentiate different instances of the same type of the template.
+In addition to the template language tag, sometimes it is preferred to localize the text of the template rather the whole template.
+
+For example, `auth_ui_login.html` defines the HTML structure and is used for all languages. What the developer wants to localize is the text.
+
+A special function named `localize` can be used to format a localized string.
+
+```html
+<input type="password" placeholder="{{ localize "enter.password" }}">
+<!-- <input type="password placeholder="Enter Password">
+```
+
+```html
+<p>{{ localize "email.sent" .email .name }}</p>
+<!-- <p>Hi John, an email has been sent to john.doe@example.com</p -->
+```
+
+`localized` takes a translation key, followed any arguments required by that translation key. If the key is not found, the key itself is returned.
+
+The translation file itself is a template so it is loaded as template as usual. For example:
+
+```
+templates/auth_ui_translation.json
+templates/zh-Hant/auth_ui_translation.json
+templates/zh-Hans/auth_ui_translation.json
+templates/ja/auth_ui_translation.json
+```
+
+The translation file is simply a flat JSON object with string keys and string values. The value is in ICU MessageFormat. Not all ICU MessageFormat arguments are supported. The supported are `select`, `plural` and `selectordinal`.
```

Comment:
I remember we have fallback mechanism for a whole template (i.e. if `/zh-Hant/template.html` does not exists, fallback to `/template.html`), but do we want to have fallback mechanism per-key and per-language?
- e.g. if `email.sent` is not found in `/zh-Hant/translation.json`, fall back to `/translation.json`
- e.g. if `email.sent` is not found in `/zh-Hant/translation.json`, fall back to `/zh/translation.json`

Reply:
Good point! Translation resolution should be different from template resolution. Let me clarify it.